### PR TITLE
Feat(remote access) using Json Web Token to handle the remote access

### LINKED ISSUE #6675: Allow remote access to bhima
As bhima will be used a lot for the next project, I think allowing other servers or applications to communicate with it will be very important. The remote access will make it easy to build other applications on top of bhima (having bhima as server for instance as DHIS2 is).

We currently need a session to access bhima APIs (private ones). It will be possible to use a <a href="https://jwt.io">token</a> for creating a session in bhima, so all APIs will be called by any application(mobile, nodes, php, etc...).

The token will not replace the session, but will be used only for remote access, it means that the current behavior will not be affected.

### PR DESCRIPTION
**Pull request by jeremielodi**
closes #6675

This snippet below describes how you can call bhima private APIs remotely:
```js

const axios = require('axios').default;
const server = 'http://localhost:8080';

axios.defaults.xsrfCookieName = 'csrftoken';
axios.defaults.xsrfHeaderName = 'X-CSRFToken';

async function getToken() {
  const endPoint = `${server}/auth/login`;
  const response = await axios.post(endPoint, {
    username: 'superuser',
    password: 'superuser',
    project: 1
  }, {
    responseType: 'json',
  });
  return response.data.token;
}

async function getDepots() {
  const endPoint = `${server}/depots`;
  const token = await getToken();
  const response = await axios.
    get(endPoint, {
      headers: {
        'x-access-token' : token,
      }
    });
  const depots = response.data;
  console.log(depots)
}

getDepots();

```

You can try the same thing using python

```python

import requests
import json


server = 'http://localhost:8080'
authAPI = server + '/auth/login'

# headers
headers = {
  'response-type': 'json',
}
# bhima requires the username, password and the project id for authentification
authData = {
  'username': 'superuser',
  'password': 'superuser',
  'project': 1
}
# let get the token for remote access
authResponse = requests.request(
  "POST", authAPI, headers=headers, json=authData
)

session = json.loads(authResponse.content)
token = session['token']

# accessing private route

depotAPI = server + '/depots'

# headers
headers = {
  'x-access-token': token,
  'response-type': 'json',
}

depotResponse = requests.request("GET", depotAPI, headers=headers)
depots = json.loads(depotResponse.content)
# let print the result
print(depots)

```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -124,6 +124,7 @@
     "jquery": "^3.5.1",
     "jsbarcode": "^3.11.5",
     "json-2-csv": "^3.17.1",
+    "jsonwebtoken": "^8.5.1",
```
Comment (jmcameron) [inline package.json:127]: I think you need to add axios too.

File: server/config/access.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (sessionExists) {
+      if (_.isUndefined(req.session.user) && !within(req.path, publicRoutes) && !withinAsset(req.path, assetRoutes)) {
+        debug(`Rejecting unauthorized access to ${req.path} from ${req.ip}`);
+        next(new Unauthorized('You are not logged into the system.'));
+      } else {
+        next();
+      }
+    } else {
+      debug('token not valide');
```
Comment (jmcameron) [inline server/config/access.js]: Spelling error (English):  valide -> valid

File: server/config/access.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+const _ = require('lodash');
```
Comment (jmcameron) [inline server/config/access.js:1]: I like the reorganization of these functions into the access.js file.

File: test/integration/remoteAccess.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
+/* global expect, chai, */
```
Comment (jmcameron) [inline test/integration/remoteAccess.js:1]: Thanks for adding a test!

**Review by jmcameron — COMMENTED** [review]
I looked over the code and it looks good.  
Thanks for adding a test!  Excellent!

I tested this and it works fine.  But axios needs to be added to the package list.

The test script should also log off!  And any applications that use this interface should log off when they complete their task.

One concern I had was that for our production sites we connect with http, not https, so the user credentials will be transmitted in the clear.  Since our regular web interface login has the same problem, this is not any less secure.  However, this is an issue that we may want to consider addressing in the future:  setting up and require https even in local LAN installations.

One comment:  I noticed that the token created this way is good for 24 hours.  That is probably the default for web interfaces.  However for this application, I wonder if a smaller period could be used.  If a program is accessing a Bhima site, wouldn't it do its operations fairly quickly and then log off?  In which case a shorter period could be used.

Finally, although I tested this and it seems to work well, I'm not familiar enough with the login authentication security implications and whether this might create future security problems.  So hopefully someone else will be able to review it and provide additional insight into the security issues.

File: package.json (hunk 1 above)
Comment (jeremielodi) [reply package.json:127]: I used ```axios``` just for snippet I provided in comment. Any node request module can be used for testing the remote access.

File: server/config/access.js (hunk 2 above)
Comment (jeremielodi) [reply server/config/access.js]: You're right

**Comment by mbayopanda** [comment]
It works like a charm 👍🏽

**Review by mbayopanda — APPROVED** [review]
LGTM 👍🏽

**Pull request merged**